nomos-services/blend/src/epoch.rs

```diff
@@ -0,0 +1,285 @@
+use core::{
+    fmt::Debug,
+    marker::PhantomData,
+    pin::Pin,
+    task::{Context, Poll},
+};
+
+use async_trait::async_trait;
+use cryptarchia_engine::{Epoch, Slot};
+use futures::{FutureExt as _, Stream, StreamExt as _};
+use nomos_core::crypto::ZkHash;
+use nomos_ledger::EpochState;
+use nomos_time::SlotTick;
+use overwatch::overwatch::OverwatchHandle;
+
+const LOG_TARGET: &str = "blend::service::epoch";
+
+#[derive(Debug, PartialEq, Eq)]
+pub struct EpochInfo {
+    nonce: ZkHash,
+    ledger_aged: ZkHash,
+    total_stake: u64,
+}
+
+impl From<EpochState> for EpochInfo {
+    fn from(
+        EpochState {
+            total_stake,
+            utxos,
+            nonce,
+            ..
+        }: EpochState,
+    ) -> Self {
+        Self {
+            ledger_aged: utxos.root(),
+            nonce,
+            total_stake,
+        }
+    }
+}
+
+/// A trait that provides the needed functionalities for the epoch stream to
+/// fetch the epoch state for a given slot.
+#[async_trait]
+pub trait ChainApi<RuntimeServiceId> {
+    async fn new(overwatch_handle: &OverwatchHandle<RuntimeServiceId>) -> Self;
+    async fn get_epoch_state_for_slot(&self, slot: Slot) -> Option<EpochState>;
+}
+
+type GetEpochForSlotFuture = Pin<Box<dyn Future<Output = Option<EpochState>> + Send>>;
+
+/// A stream that listens to slot ticks, and on the first slot tick received as
+/// well as the first slot tick of each new epoch, fetches the epoch state from
+/// the provided chain service adapter.
+///
+/// In case the epoch state for a given slot is not found, it will retry on
+/// subsequent slots until one is successfully received.
+pub struct EpochStream<SlotStream, ChainService, RuntimeServiceId> {
+    slot_stream: SlotStream,
+    chain_service: ChainService,
+    current_epoch: Option<Epoch>,
+    epoch_state_future: Option<GetEpochForSlotFuture>,
+    _phantom: PhantomData<RuntimeServiceId>,
+}
+
+impl<SlotStream, ChainService, RuntimeServiceId>
+    EpochStream<SlotStream, ChainService, RuntimeServiceId>
+{
+    pub fn new(slot_stream: SlotStream, chain_service: ChainService) -> Self {
+        Self {
+            slot_stream,
+            chain_service,
+            current_epoch: None,
+            epoch_state_future: None,
+            _phantom: PhantomData,
+        }
+    }
+}
+
+#[expect(
+    clippy::cognitive_complexity,
+    reason = "It's better to keep all the polling logic together."
+)]
+impl<SlotStream, ChainService, RuntimeServiceId> Stream
```

Comment:
I have the feeling this could be simplified using stream operations instead of manual operations.

Reply:
I mean, the logic is a bit too much complicated to be performed entirely with stream operations. I thought about it, but arrived to the conclusion this would be actually easier to read. We poll the first time and on each epoch change, and on the first slot we fetch the epoch state, and retry on the next slot in case the previous attempt failed. I could not find an easy enough way to express this without a custom `Stream` implementation.

Reply:
Why not transform the slot_stream to an epoch stream? Then it should be simpler?
I trust your decision tho. If you say its better you know better.

Reply:
Yes, but we still need to tick on every slot, since unfortunately it's possible that `GetEpochState` returns `None` for a given slot, in which case I decided that we should try on the next slot tick again, instead of panicking. I can't think of a way to implement this logic in a more idiomatic way (?)